runtime/tests/jvmtitests/src/com/ibm/jvmti/tests/fieldwatch/fw001.c

```diff
@@ -0,0 +1,233 @@
+			printf("Field Access Report #%d: method: %p location %d. Class: %p, fieldID: %p\n", numReports, method, (jint)location, field_klass, field);
+		}
+		else {
+			printf("**Error**: Invalid value for numReports. Return Code: %d\n", numReports);
+		}
+	}
+}
+
+/* Callback routine invoked by the VM to report a field write event. */
+void JNICALL
+fieldModificationCallback(jvmtiEnv *jvmti_env, JNIEnv* jni_env, jthread thread, jmethodID method, jlocation location, jclass field_klass, jobject object, jfieldID field, char signature_type, jvalue new_value)
+{
+	/* Only increment the count if the field ID matches with the same field we added the watch for. */
+	if (currentFieldBeingWatched == field) {
+		jint numReports = incAndGetNumReports(jni_env);
+		if (numReports >= JNI_ERR) {
+			printf("Field Modification Report #%d: method: %p location: %d, fieldID: %p, signature: %c\n", numReports, method, (jint)location, field, signature_type);
+		}
+		else {
+			printf("**ERROR**: Invalid value for numReports. Return Code: %d\n", numReports);
+		}
+	}
+}
+
+jint JNICALL
+fw001(agentEnv * agent_env, char * args)
+{
+	jvmtiCapabilities capabilities;
+	jvmtiEventCallbacks callbacks;
+	jvmtiError err;
+	JavaVM *vm = agent_env->vm;
+	jint rc = (*vm)->GetEnv(vm, (void **) &globalEnv, JVMTI_VERSION_1_2);
+
+	if (JNI_OK != rc) {
+		printf("**ERROR**: Failed to get agent env. Return code: %d.\n", rc);
+		return JNI_ERR;
+	}
+
+	memset(&capabilities, 0, sizeof(jvmtiCapabilities));
+	capabilities.can_generate_field_access_events = 1;  
+	capabilities.can_generate_field_modification_events = 1;
+	err = (*globalEnv)->AddCapabilities(globalEnv, &capabilities);
+	if (JVMTI_ERROR_NONE != err) {
+		printf("**ERROR**: Failed to AddCapabilities. JVMTI_ERROR code: %d.\n", err);
+		return JNI_ERR;
+	}
+
+	memset(&callbacks, 0, sizeof(jvmtiEventCallbacks));
+	callbacks.FieldAccess = fieldAccessCallback;
+	callbacks.FieldModification = fieldModificationCallback;
+	err = (*globalEnv)->SetEventCallbacks(globalEnv, &callbacks, sizeof(jvmtiEventCallbacks));
+	if (JVMTI_ERROR_NONE != err) {
+		printf("**ERROR**: Failed to set callbacks. JVMTI_ERROR code:  %d.\n", err);
+		return JNI_ERR;
+	}
+
+	err = (*globalEnv)->SetEventNotificationMode(globalEnv, JVMTI_ENABLE, JVMTI_EVENT_FIELD_ACCESS, NULL);
+	if (JVMTI_ERROR_NONE != err) {
+		printf("**ERROR**: Failed to enable field access event. JVMTI_ERROR code: %d.\n", err);
+		return JNI_ERR;
+	}
+
+	err = (*globalEnv)->SetEventNotificationMode(globalEnv, JVMTI_ENABLE, JVMTI_EVENT_FIELD_MODIFICATION, NULL);
+	if (JVMTI_ERROR_NONE != err) {
+		printf("**ERROR**: Failed to enable field modification event. JVMTI_ERROR code: %d.\n", err);
+		return JNI_ERR;
+	}
+
+	return JNI_OK;
+}
+
+/* Add class name to a global so we can invoke the appropriate Java API to update the number of reports seen. */
+JNIEXPORT jint JNICALL
+Java_com_ibm_jvmti_tests_fieldwatch_fw001_startTest(JNIEnv *jni_env, jobject cls, jstring name)
+{
+	testClassString = name;
```

Comment:
This is not legal - if you want to store a JNI ref to use outside of the scope in which it was created, you need to make it into a global ref (search for NewGlobalRef). It should then be deleted in endTest (DeleteGlobalRef).

A much better solution would be to pass the String again to endTest, then there's no need to store it at all.